### PR TITLE
Firebase data tables on by default

### PR DESCRIPTION
**Pull request by shans**
This turned out much harder to do than I'd hoped, in particular it led to a bunch of issues uncovering other issues.

For example, I discovered that we were creating duplicates of all create/copy stores, giving the original to the first requesting particle and the duplicate to all subsequent particles. This *forced* almost all arcs communication through the slow path rather than in-security-domain.

Fixing this showed that we were also creating multiple duplicates of the special User / Users / etc. stores.

Once that was fixed I found out that the storage proxy classes didn't cover a bunch of corner cases, because we'd never put them in these cases, because there were always 2 of them running in parallel.

etc...

At any rate, this is not 100% working. All the unit tests pass, and hopefully the webdriver ones will too. The demos *mostly* work, but there are 2 issues:
(1) sometimes when you share an arc it doesn't quite load correctly the first time; if this happens you can refresh and then it does load. I need to look into this further and would prefer to do so once the patch is merged so that I'm not falling further and further behind head.
(2) the duplicate slot thing is back, when a friend adds a favourite food to a shared arc. I was hoping Scott could take a look at this as he's fixed similar issues in the past.

### CODE REVIEW

File: runtime/ts/storage/crdt-collection-model.ts [hunk 1]
```diff
@@ -43,6 +43,7 @@ export class CrdtCollectionModel {
   // or `value` is different to the value previously stored).
   add(id: string, value, keys): boolean {
     assert(keys.length > 0, 'add requires keys');
+    assert(typeof keys === 'object', 'add requires a list of keys');
```
Comment (mykmartin) [inline runtime/ts/storage/crdt-collection-model.ts]: Why is this required? Does the first assert not suffice?

File: runtime/ts/storage/in-memory-storage.ts [hunk 2]
```diff
@@ -155,22 +148,39 @@ abstract class InMemoryStorageProvider extends StorageProviderBase {
     }
     return new InMemoryVariable(type, storageEngine, name, id, key);
   }
+
+  async ensureBackingStore() {
+    if (this.backingStore) {
+      return this.backingStore;
```
Comment (mykmartin) [inline runtime/ts/storage/in-memory-storage.ts]: Missing await?  (Or, since we know backingStore is a non-Promise value, does the other instance have an unnecessary await?)

File: runtime/ts/storage/in-memory-storage.ts [hunk 3]
```diff
@@ -155,22 +148,39 @@ abstract class InMemoryStorageProvider extends StorageProviderBase {
     }
     return new InMemoryVariable(type, storageEngine, name, id, key);
   }
+
+  async ensureBackingStore() {
```
Comment (mykmartin) [inline runtime/ts/storage/in-memory-storage.ts]: It seems most call sites await (or .then) the return from this - do we need to have the internal awaits as well?

File: runtime/ts/storage/in-memory-storage.ts [hunk 4]
```diff
@@ -155,22 +148,39 @@ abstract class InMemoryStorageProvider extends StorageProviderBase {
...
+    }
+    if (!this.pendingBackingStore) {
+      const key = this.storageEngine.baseStorageKey(this.backingType());
+      this.pendingBackingStore = this.storageEngine.baseStorageFor(this.type, key);
+      this.backingStore = await this.pendingBackingStore;
+    } else {
+      await this.pendingBackingStore;
+    }
+    return this.backingStore;
+  }
```
Comment (mykmartin) [inline runtime/ts/storage/in-memory-storage.ts]: Possible simplification:
```
async ensureBackingStore() {
  if (!this.backingStore) {
    if (!this.pendingBackingStore) {
      const key = this.storageEngine.baseStorageKey(this.backingType());
      this.pendingBackingStore = this.storageEngine.baseStorageFor(this.type, key);
    }
    this.backingStore = await this.pendingBackingStore;
    this.pendingBackingStore = null;
  }
  return [await?] this.backingStore;
}
```

File: runtime/ts/storage/firebase-storage.ts [hunk 5]
```diff
@@ -176,26 +179,40 @@ export class FirebaseStorage extends StorageBase {
...
       }
+      enableReferenceMode = referenceMode;
     }
 
-    return FirebaseStorageProvider.newProvider(type, this, id, reference, key, shouldExist);
+<<<<<<< HEAD
+    let provider = FirebaseStorageProvider.newProvider(type, this, id, reference, fbKey, shouldExist);
+=======
+    const provider = FirebaseStorageProvider.newProvider(type, this, id, reference, key, shouldExist);
+>>>>>>> lint & tslint fixes
```
Comment (mykmartin) [inline runtime/ts/storage/firebase-storage.ts]: :)

File: runtime/ts/storage/firebase-storage.ts [hunk 6]
```diff
@@ -215,19 +232,37 @@ abstract class FirebaseStorageProvider extends StorageProviderBase {
...
     this.reference = reference;
     this.backingStore = null;
+    this.pendingBackingStore = null;
 
     // Resolved when local modifications complete being persisted
     // to firebase. Null when not persisting.
     this.persisting = null;
   }
 
+  async ensureBackingStore(arg='') {
```
Comment (mykmartin) [inline runtime/ts/storage/firebase-storage.ts]: Left-over debugging artifact (arg)

File: runtime/ts/storage/firebase-storage.ts [hunk 7]
```diff
@@ -342,53 +380,91 @@ class FirebaseVariable extends FirebaseStorageProvider {
...
     if (this.localModified) {
       return;
     }
     const data = dataSnapshot.val();
     assert(this.version == null || data.version > this.version);
 
-    this.value = data.value;
+    // NOTE that remoteStateChanged will be invoked immediately by the
+    // this.reference.on(...) call in the constructor; this means that it's possible for this
+    // function to recieve data with storageKeys before referenceMode has been switched on (as
```
Comment (mykmartin) [inline runtime/ts/storage/firebase-storage.ts]: Typo: s/recieve/receive

File: runtime/ts/storage/firebase-storage.ts [hunk 8]
```diff
@@ -397,67 +473,93 @@ class FirebaseVariable extends FirebaseStorageProvider {
...
+      this.value = value;
     }
-
     this.localModified = true;
-    this.value = value;
-    this._fire('change', {data: this.value, version: this.version, originatorId, barrier});
-    await this._persistChanges();
+
+    // the await required for fetching baseStorage can cause initialization/localModified
+    // flag reordering if done before persisting a change.
```
Comment (mykmartin) [inline runtime/ts/storage/firebase-storage.ts]: ..., so we need to fire the change event after persistChanges has finished?

File: shell/test/specs/demo-tests.js [hunk 9]
```diff
@@ -181,7 +181,7 @@ function waitForStillness() {
     }
     return matches > desiredMatches;
   };
-  browser.waitUntil(noGlow, 20000, `the glow can't stop won't stop`, 500);
```
Comment (shaper) [inline shell/test/specs/demo-tests.js]: Do we want to check in this boost here (and below?)

File: runtime/arc.js [hunk 10]
```diff
@@ -435,16 +436,27 @@ ${this.activeRecipe.toString()}`;
         recipeHandle.id = newStore.id;
         recipeHandle.fate = 'use';
         recipeHandle.storageKey = newStore.storageKey;
+        continue;
         // TODO: move the call to ParticleExecutionHost's DefineHandle to here
       }
 
+      // TODO(shans/sjmiles): This shouldn't be possible, but at the moment the
+      // shell pre-populates all arcs with a set of handles so if a recipe explicitly
+      // asks for one of these there's a conflict.
```
Comment (shaper) [inline runtime/arc.js]: Helpful to state briefly here what we want to happen eventually instead?

File: runtime/arc.js [hunk 11]
```diff
@@ -174,7 +174,8 @@ export class Arc {
             context.dataResources.set(storageKey, storeId);
             // TODO: can't just reach into the store for the backing Store like this, should be an 
             // accessor that loads-on-demand in the storage objects.
-            await this._serializeHandle(handle._backingStore, context, storeId);
+            await handle.ensureBackingStore();
```
Comment (shaper) [inline runtime/arc.js]: Here and the changes below in this file, are there unit tests that would be useful to add to cover these changes?

I'd infer that we had to make these changes because other work in this PR caused existing tests to fail. So in that sense they are covered, but in another sense they were broken previously and we didn't know, so maybe more granular tests for the specific cases would be helpful.

I am ok if you think it's fine as is, wanted to ask while reading.

File: runtime/manifest.js [hunk 12]
```diff
@@ -877,6 +877,7 @@ ${e.message}
           const hostedParticleLiteral = hostedParticle.clone().toLiteral();
           let particleSpecHash = await digest(JSON.stringify(hostedParticleLiteral));
           let id = `${manifest.generateID()}:${particleSpecHash}:${hostedParticle.name}`;
+          hostedParticleLiteral.id = id;
```
Comment (shaper) [inline runtime/manifest.js]: We didn't have this and we had to add it for some reason, either a comment or a test or both?

File: runtime/storage-proxy.js [hunk 13]
```diff
@@ -157,8 +159,24 @@ class StorageProxyBase {
   }
 
   _processUpdates() {
+
+    let updateIsNext = update => {
+      if (update.version == this._version + 1) {
+        return true;
+      }
+      // Holy Layering Violation Batman
```
Comment (shaper) [inline runtime/storage-proxy.js]: Add a brief comment on whether this is expected to be here long term or if there's other work planned to improve?

File: runtime/storage-proxy.js [hunk 14]
```diff
@@ -352,6 +377,17 @@ class VariableProxy extends StorageProxyBase {
     if (this._barrier != null) {
       if (update.barrier == this._barrier) {
         this._barrier = null;
+
+        // HOLY LAYERING VIOLATION BATMAN
```
Comment (shaper) [inline runtime/storage-proxy.js]: Similar to above comment -- is this long term or will it go away at some point, if so, what will drive that.

**Comment by shaper** [comment]
Also -- file issues for (1) and (2) mentioned in the PR comment? Could help w/ tracking and discussion if they don't already exist.

My questions around unit tests in my review just now were in part motivated by your PR description. It sounds like we uncovered some hidden issues and added fixes to get tests to pass but it doesn't look like we added new more granular tests that would have *just* caught those hidden issues. Maybe that's fine but I wanted to ask since now would be the time to consider to get ahead of similar antics in the future.

File: shell/test/specs/demo-tests.js (hunk 9 above)
Comment (shans) [reply shell/test/specs/demo-tests.js]: No, I'm trying to debug the wdio tests; I can't launch them locally, and on Michael's machine they run to completion. Timing out was one hypothesis.

File: runtime/arc.js (hunk 11 above)
Comment (shans) [reply runtime/arc.js]: Yup, these changes are related to existing test failures. Most of these tests started failing because of the situation I described in the headline comment - duplicated handles leading to a lack of coverage of concurrency mechanisms. In that sense, they were broken previously because of an unrelated component's failure. This is obviously concerning in other ways - how many other tests don't currently test what they are supposed to be testing?

I would like to add a bunch more tests around the channel, but I need to work out the right way to do this and I think that the coverage for this patch is probably sufficient.

File: runtime/ts/storage/crdt-collection-model.ts (hunk 1 above)
Comment (shans) [reply runtime/ts/storage/crdt-collection-model.ts]: this might just be a merge duplication? Not sure that the second assert does add anything.

File: runtime/ts/storage/in-memory-storage.ts (hunk 4 above)
Comment (mykmartin) [reply runtime/ts/storage/in-memory-storage.ts]: Looking at this again, it might not be correct to set pendingBackingStore to null after the await, since that could cause another invocation of ensureBackingStore awaiting at the same point to then write null to backingStore?

File: runtime/ts/storage/crdt-collection-model.ts (hunk 1 above)
Comment (shans) [reply runtime/ts/storage/crdt-collection-model.ts]: as you demonstrated to me, it's so that single strings aren't accepted (these also have a length).

File: runtime/ts/storage/in-memory-storage.ts (hunk 2 above)
Comment (shans) [reply runtime/ts/storage/in-memory-storage.ts]: Figured out how to get rid of the extra await :)

File: runtime/ts/storage/in-memory-storage.ts (hunk 3 above)
Comment (shans) [reply runtime/ts/storage/in-memory-storage.ts]: There's one asynchronous requirement - this.backingStore needs to await the output of this.storageEngine.baseStorageFor. However, that can happen in a `then()`, which means we can directly return this.pendingBackingStore. Awaiting this function guarantees that this.backingStore will exist afterwards, as the `then()` assignment must execute before anything scheduled after that await.

File: runtime/ts/storage/in-memory-storage.ts (hunk 4 above)
Comment (shans) [reply runtime/ts/storage/in-memory-storage.ts]: Yeah, I definitely feel like I'm sinking here.

File: runtime/ts/storage/firebase-storage.ts (hunk 6 above)
Comment (shans) [reply runtime/ts/storage/firebase-storage.ts]: removed.

File: runtime/ts/storage/firebase-storage.ts (hunk 7 above)
Comment (shans) [reply runtime/ts/storage/firebase-storage.ts]: fixed.

File: runtime/ts/storage/firebase-storage.ts (hunk 8 above)
Comment (shans) [reply runtime/ts/storage/firebase-storage.ts]: oops, out-of-date comment.

File: runtime/arc.js (hunk 10 above)
Comment (shans) [reply runtime/arc.js]: yup, thanks. Done.

File: runtime/manifest.js (hunk 12 above)
Comment (shans) [reply runtime/manifest.js]: A comment inline there would be weird. Everything we store should have an ID. We got away with not having this for stored particles before but once they can be stored by reference then a bunch of tests break if they don't have one.

File: runtime/ts/storage/in-memory-storage.ts [hunk 15]
```diff
@@ -155,22 +148,41 @@ abstract class InMemoryStorageProvider extends StorageProviderBase {
...
+  async ensureBackingStore() {
+    if (this.backingStore) {
+      return this.backingStore;
+    }
+    if (!this.pendingBackingStore) {
+      const key = this.storageEngine.baseStorageKey(this.backingType());
+      this.pendingBackingStore = this.storageEngine.baseStorageFor(this.type, key);
+      this.pendingBackingStore.then(backingStore => this.backingStore = backingStore);
+    }
+    return this.pendingBackingStore;
```
Comment (mykmartin) [inline runtime/ts/storage/in-memory-storage.ts]: Nice rework. One possible concern - does this mean the pendingBackingStorePromise will always hang around after the backing store has been resolved? Not a huge problem, but might be good to discard the memory for it.

**Comment by shans** [comment]
Issue (1) seems to have been resolved by a fix I needed to make to make a test pass that was only showing up as failures on the bots (see d2a0595 - "Add missing await").

**Comment by shans** [comment]
This causes https://github.com/PolymerLabs/arcs/issues/1852 and https://github.com/PolymerLabs/arcs/issues/1853 but as discussed I think it's worth landing anyway, sorry about generating more bugs :(
